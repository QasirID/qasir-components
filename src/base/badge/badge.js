import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import * as c from '../../shared/constants/color';
import * as v from '../../shared/constants/variables';

export const BADGE_VARIANT = {
  alert: 'alert',
  positive: 'positive',
  warning: 'warning',
  discount: 'discount'
};

class Badge extends PureComponent {
  getStatus(status) {
    switch (status) {
      case BADGE_VARIANT.alert:
        return css`
          background: ${c.RedPrimary};
          color: ${c.LightPrimary};
        `;
      case BADGE_VARIANT.positive:
        return css`
          background: ${c.GreenPrimary1};
          color: ${c.LightPrimary};
        `;
      case BADGE_VARIANT.warning:
        return css`
          background: ${c.YellowPrimary};
        `;
      case BADGE_VARIANT.discount:
        return css`
          color: ${c.LightPrimary};
          background: ${c.OrangePrimary};
        `;
      default:
        return false;
    }
  }

  render() {
    const { content, status, isFloating, ...props } = this.props;
    const multipleWord = Object.keys(content).length > 0 ? 'multiple-word' : undefined;

    const StyledBadge = styled.div`
      margin: 0;
      display: inline-block;
      font-size: ${v.FontSize.tiny}px;
      min-width: 20px;
      height: 20px;
      line-height: 20px;
      white-space: nowrap;
      padding: 0 6px;
      border-radius: 10px;
      box-shadow: 0 0 0 2px ${c.LightPrimary};
      ${this.getStatus(status)}

      &.multiple-word {
        padding: 0 ${v.spacing.xs}px;
      }
    `;

    const BadgeContainer = styled.div`
      position: relative;
      z-index: 10;

      &.floating {
        position: absolute;
        right: -${v.spacing.xs}px;
        top: -${v.spacing.xs}px;
      }
    `;

    return (
      <BadgeContainer
        className={isFloating ? 'floating' : undefined}
        {...props}
      >
        <StyledBadge className={multipleWord}>{content}</StyledBadge>
      </BadgeContainer>
    );
  }
}

Badge.propTypes = {
  /**
   * Content of badge, can be string or number
   */
  content: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Status of badge, Must be one of these: `alert`,`warning` or `positive`.
   */
  status: PropTypes.oneOf(
    Object.keys(BADGE_VARIANT).map(type => BADGE_VARIANT[type])
  ),
  /**
   * Specify badge floating position
   */
  isFloating: PropTypes.bool
};

Badge.defaultProps = {
  content: undefined,
  status: 'alert',
  isFloating: false
};

export default Badge;
