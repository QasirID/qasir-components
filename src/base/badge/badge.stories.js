import React from 'react';
import { storiesOf } from '@storybook/react';

import { text, select, boolean } from '@storybook/addon-knobs';
import Badge, { BADGE_VARIANT } from './badge';

/**
 * Put here:
 * - description
 */
const DESCRIPTION = `
Badge
---
~~~js
import Badge from 'qasir/lib/base';
~~~

or
~~~js
import Badge from 'qasir/lib/base/badge';
~~~

Usage
~~~js
<Badge />
~~~
`;

storiesOf('base|Badge', module)
  .addParameters({
    info: {
      text: DESCRIPTION
    }
  })
  .add('all variants', () => (
    <>
      <h3>Example with Text</h3>
      <div style={{ display: 'flex', flexDirection: 'row' }}>
        {Object.values(BADGE_VARIANT).map(variant => (
          <div
            key={variant}
            style={{
              marginRight: 5
            }}
          >
            <Badge content={variant} status={variant} />
          </div>
        ))}
      </div>
      <h3>Example with Number</h3>
      <div style={{ display: 'flex', flexDirection: 'row' }}>
        {Object.values(BADGE_VARIANT).map(variant => (
          <div
            key={variant}
            style={{
              marginRight: 5
            }}
          >
            <Badge content={5} />
          </div>
        ))}
      </div>

      <h3>Floating Example</h3>
      <div
        style={{
          width: 100,
          height: 100,
          background: '#EEE',
          borderRadius: 4,
          position: 'relative'
        }}
      >
        <Badge content={2} isFloating />
      </div>
    </>
  ))
  .add('with knobs', () => (
    <>
      <div
        style={{
          width: 100,
          height: 100,
          background: '#EEE',
          borderRadius: 4,
          position: 'relative'
        }}
      >
        <Badge
          content={text('content', 2)}
          status={select('status', BADGE_VARIANT, 'alert')}
          isFloating={boolean('isFloating', false)}
        />
      </div>
    </>
  ));
