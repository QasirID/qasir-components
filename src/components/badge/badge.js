import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { darken } from 'polished';

import CONSTANTS from '../../modifiers/constants';
import colors from '../../modifiers/colors';
// import { button } from '../../modifiers/variables';

const BadgeContainer = styled.div`
    position: relative;
`;

const StyledBadge = styled.div
        .attrs(({ disabled }) => ({ 
            disabled: disabled,
            className: 'qasir-badge'
        }))`
    border          : 1px solid #fff;
    border-radius   : 32px;
    padding         : ${ props => props.dot ? '4px' : '2px 4px' };
    margin          : 0;
    font-size       : 10px;
    font-weight     : bold;
    color           : #fff;
    background-color: ${ props => props.color ? props.color : colors.red };
    display         : inline-block;
    position        : absolute;
    left            : ${ props => props.dot ? 'calc(100% - 6px)' : 'calc(100% - 12px)' };
    top             : ${ props => props.dot ? '-4px' : '-8px'};
    z-index         : 9;
`;

export const Badge = ({ count, dot, color, showWhenZero }) => {
    return (
        <BadgeContainer>
            { (showWhenZero || count > 0) &&
                <StyledBadge dot={dot} color={color}>{ !dot && ( count > 99 ? '99+' : count ) }</StyledBadge>
            }
        </BadgeContainer>);
}

Badge.propTypes = {
    count: PropTypes.number,
    color: PropTypes.string,
    dot: PropTypes.bool,
    showWhenZero: PropTypes.bool,
}

Badge.defaultProps = {
    count: 0,
    color: colors.red,
    dot: false,
    showWhenZero: false,
}

export default Badge;
