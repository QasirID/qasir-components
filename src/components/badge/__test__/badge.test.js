import React from "react";
import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";
import { Badge } from "../badge";

describe('Badge component snapshot', () => {
    test ('should match', () => {
        const component = renderer.create(<Badge count={38}  />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('Badge', () => {
    test("should match text with count", () => {
        const badge = shallow(<Badge count={38} />);
        expect(badge.text()).toEqual('38');
    });
    
    test("should not match text with count, text should not be displayed", () => {
        const badge = shallow(<Badge count={38} dot={true} />);
        expect(badge.text()).toEqual('');
    });

    test("should hide when count is 0", () => {
        const badge = mount(<Badge count={0} />);
        expect(badge.find('.qasir-badge').exists()).toEqual(false);
    });

    test("should be shown when count is 0", () => {
        const badge = mount(<Badge count={0} showWhenZero={true} />);
        expect(badge.find('.qasir-badge').exists()).toEqual(true);
    });
});
