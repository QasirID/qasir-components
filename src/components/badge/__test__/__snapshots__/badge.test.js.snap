// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge component snapshot should match 1`] = `
<div
  className="sc-bdVaJa jaOyYS"
>
  <div
    className="sc-bwzfXH qasir-badge fQeyyG"
    color="#F04B32"
  >
    38
  </div>
</div>
`;
